src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,267 @@
+package jdk.incubator.code.internal;
+
+import com.sun.tools.javac.code.*;
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.TreeMaker;
+import com.sun.tools.javac.util.*;
+import jdk.incubator.code.*;
+import jdk.incubator.code.op.CoreOp;
+import jdk.incubator.code.op.ExternalizableOp;
+import jdk.incubator.code.op.OpFactory;
+import jdk.incubator.code.type.*;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.sun.tools.javac.code.Flags.*;
+
+public class CodeModelToAST {
+
+    private final TreeMaker treeMaker;
+    private final Names names;
+    private final Symtab syms;
+    private final Symbol.ClassSymbol currClassSym;
+    private final CodeReflectionSymbols crSym;
+    private final Map<Value, JCTree> valueToTree = new HashMap<>();
+    private final Map<JavaType, Type> jtToType;
+    private Symbol.MethodSymbol ms;
+
+    public CodeModelToAST(TreeMaker treeMaker, Names names, Symtab syms,
+                          Symbol.ClassSymbol currClassSym, CodeReflectionSymbols crSym) {
+        this.treeMaker = treeMaker;
+        this.names = names;
+        this.syms = syms;
+        this.currClassSym = currClassSym;
+        this.crSym = crSym;
+        this.jtToType = mappingFromJavaTypeToType();
+    }
+
+    private Map<JavaType, Type> mappingFromJavaTypeToType() {
+        Map<JavaType, Type> m = new HashMap<>();
+        Symbol.ModuleSymbol jdk_incubator_code = syms.enterModule(names.jdk_incubator_code);
```

Comment:
This is effectively an extension to javac's symbol table. My preference would be to:
* add whatever symbol/type you need in `CodeReflectionSymbol`
* then set up the translation map simply, like you do for primitive types -- e.g.
```
Map.entry(<type element>, <javac type>)
```